docs/maintain-a-collection-of-new-products/script.liquid

```diff
@@ -1,115 +1,197 @@
+  {% assign qualifying_product_ids = qualifying_product_ids | concat: result_product_ids %}
+
+  {% if result.data.products.pageInfo.hasNextPage %}
+    {% assign cursor = result.data.products.edges.last.cursor %}
+  {% else %}
+    {% break %}
+  {% endif %}
+{% endfor %}
+
+{% assign current_product_ids =  collection.products | map: "admin_graphql_api_id" %}
+
+{% assign product_ids_to_add = array %}
+{% assign product_ids_to_remove = array %}
+
+{% for qualifying_product_id in qualifying_product_ids %}
+  {% unless current_product_ids contains qualifying_product_id %}
+    {% assign product_ids_to_add = product_ids_to_add | push: qualifying_product_id %}
+  {% endunless %}
+{% endfor %}
+
+{% for current_product_id in current_product_ids %}
+  {% unless qualifying_product_ids contains current_product_id %}
+    {% assign product_ids_to_remove = product_ids_to_remove | push: current_product_id %}
+  {% endunless %}
+{% endfor %}
+
+{% assign mutations = array %}
+
+{% if product_ids_to_add != blank  %}
+  {% assign groups_of_product_ids_to_add = product_ids_to_add | in_groups_of: 250, fill_with: false %}
+
+  {% for group_of_product_ids_to_add in groups_of_product_ids_to_add %}
+    {% capture mutation %}
+      collectionAddProducts{{ forloop.index0 }}: collectionAddProducts(
+        id: {{ collection.admin_graphql_api_id | json }}
+        productIds: {{ group_of_product_ids_to_add | json }}
+      ) {
+        userErrors {
+          field
+          message
+        }
       }
-    }
+    {% endcapture %}
+
+    {% assign mutations = mutations | push: mutation %}
   {% endfor %}
+{% endif %}
 
-  {% for product in final_products %}
-    {% if product_ids_to_keep contains product.id %}
-      {% continue %}
-    {% endif %}
+{% if product_ids_to_remove != blank  %}
+  {% assign groups_of_product_ids_to_remove = product_ids_to_remove | in_groups_of: 250, fill_with: false %}
 
-    {
-      "action": {
-        "type": "shopify",
-        "options": [
-          "create",
-          "collect",
-          {
-            "collection_id": {{ collection.id | json }},
-            "product_id": {{ product.id | json }}
-          }
-        ]
+  {% for group_of_product_ids_to_remove in groups_of_product_ids_to_remove %}
+    {% capture mutation %}
+      collectionRemoveProducts{{ forloop.index0 }}: collectionRemoveProducts(
+        id: {{ collection.admin_graphql_api_id | json }}
+        productIds: {{ group_of_product_ids_to_remove | json }}
+      ) {
+        userErrors {
+          field
+          message
+        }
       }
-    }
+    {% endcapture %}
+
+    {% assign mutations = mutations | push: mutation %}
   {% endfor %}
 {% endif %}
+
+{% if mutations != blank %}
+  {% action "shopify" %}
+    mutation {
+      {{ mutations | join: newline }}
+    }
+  {% endaction %}
+{% else %}
+  {% log "This collection already has the correct number of products." %}
```

Comment:
nit: this is more about the correct _products_ being in place, rather than the correct _number_ of products, right?

Reply:
true, true, update forthcoming